### PR TITLE
Add recommendations around variables usage

### PR DESCRIPTION
**Pull request by ericzolf**
I also extended the inventory to usage of variables overall

### CODE REVIEW

File: inventories/README.adoc [hunk 1]
```diff
@@ -255,3 +256,52 @@ create some file to simulate an API call to provision a host ------------ 0.57s
 +
 TIP: if for some reason, you can't follow the recommendation, you can at least avoid duplicating too much information by indirectly referencing the hosts' variables as in `"{{ hostvars[item.name]['provision_value'] }}"`. Not so bad...
 ====
+
+== Restrict your usage of variable types
+[%collapsible]
+====
+Explanations::
+* Do not use playbook and play variables, as well as include_vars, prefer inventory variables.
```
Comment (mophahr) [inline inventories/README.adoc:264]:
```suggestion
* Avoid playbook and play variables, as well as `include_vars`. 
  Opt for inventory variables instead.
```

File: inventories/README.adoc [hunk 2]
```diff
@@ -255,3 +256,52 @@ create some file to simulate an API call to provision a host ------------ 0.57s
...
++
+. role defaults (defined in `defaults/main.yml`), they are... defaults and can be overwritten by anything.
+. inventory vars, they truly represent your desired state.
+  They have their own internal precedence (group before host) but that's easy to remember.
+. host facts don't represent a desired state but the current state, and no other variable should have the same name because of <<differentiate>> so that the precedence doesn't really matter.
+. role vars (defined in `vars/main.yml`) represent constants used by the role to separate code from data, and shouldn't either collide with the inventory variables, but can be overwritten by extra vars if you know what you're doing.
+. scoped vars, at the block or task level, are local to their scope and hence internal to the role, and can't collide with other variable types.
+. runtime vars, defined by register or set_facts, are taking precedence over almost everything defined previously, which makes sense as they represent the current state of the automation.
```
Comment (mophahr) [inline inventories/README.adoc:278]:
```suggestion
. runtime vars, defined by `register` or `set_facts`, are taking precedence over almost everything defined previously, which makes sense as they represent the current state of the automation.
```

File: inventories/README.adoc [hunk 3]
```diff
@@ -255,3 +256,52 @@ create some file to simulate an API call to provision a host ------------ 0.57s
...
+. scoped params, at the role or include level this time, are admittedly a bit out of order and should be avoided to limit surprises.
```
Comment (mophahr) [inline inventories/README.adoc:279]:
```suggestion
. scoped params, defined at the role or include level (as opposed to block or task level for scoped variables), are admittedly a bit out of order and should be avoided to limit surprises.
```

File: inventories/README.adoc [hunk 4]
```diff
@@ -255,3 +256,52 @@ create some file to simulate an API call to provision a host ------------ 0.57s
...
+. and extra_vars overwrite everything else (even runtime vars, which can be quite surprising)
```
Comment (mophahr) [inline inventories/README.adoc:280]:
```suggestion
. and lastly, extra vars overwrite everything else (even runtime vars, which can be quite surprising)
```

File: inventories/README.adoc [hunk 5]
```diff
@@ -255,3 +256,52 @@ create some file to simulate an API call to provision a host ------------ 0.57s
...
+
+The following picture shows how this can be summarized in a simplified way to keep in mind, highlighting which variables are meant to overwrite each other:
```
Comment (mophahr) [inline inventories/README.adoc:282]:
```suggestion
The following picture summarizes this list in a simplified and easier to keep in mind way, highlighting which variables are meant to overwrite others:
```

File: inventories/README.adoc [hunk 6]
```diff
@@ -255,3 +256,52 @@ create some file to simulate an API call to provision a host ------------ 0.57s
...
+
+.Flow of variable precedences
+image::variable_precedences.png[flow of variable precedences in 3 lanes]
+
+CAUTION: even if we write that variables _shouldn't_ overwrite each other, they still all share the same namespace and _can_ potentially overwrite each other.
+It is your responsibility as automation author to make sure they don't.
+====
+
+== Prefer inventory variables for desired state variables over extra vars
```
Comment (mophahr) [inline inventories/README.adoc:291]:
```suggestion
== Prefer inventory variables over extra vars to describe the desired state
```

File: inventories/README.adoc [hunk 7]
```diff
@@ -255,3 +256,52 @@ create some file to simulate an API call to provision a host ------------ 0.57s
...
+[%collapsible]
+====
+Explanations::
+Don't use extra vars to define your desired state.
+Make sure your inventory contains the complete state of your environment as you want to reach.
```
Comment (mophahr) [inline inventories/README.adoc:296]:
```suggestion
Make sure your inventory completely describes how your environment is supposed to look like.
```

File: inventories/README.adoc [hunk 8]
```diff
@@ -255,3 +256,52 @@ create some file to simulate an API call to provision a host ------------ 0.57s
...
+Use extra vars only for troubleshooting, debugging or validation purposes.
+
+Rationale::
+Inventory variables are typically in some kind of persistent tracked storage (be it a database or Git), and should be your sole source representing your desired state so that you can refer to it non-ambiguously.
+On the other hand, extra vars are bound to a specific job or ansible-call and disappear together with history. 
+
+Examples::
```
Comment (mophahr) [inline inventories/README.adoc:303]: Maybe add `no_log: "{{ overwrite_my_by_extra_var | default(true) }}"` as an example

**Review by mophahr — COMMENTED** [review]
No objections on a content level.

But I struggled a bit with some of the wordings, so I left some suggestions that, in my opinion, make it more readable.

File: inventories/README.adoc (hunk 1 above)
Comment (ericzolf) [reply inventories/README.adoc:264]: I won't comment on all your... comments but I'm taking them over in one big commit, and not one by one. I'm resolving them as I include them (without further... comment).

**Comment by ericzolf** [comment]
Ready for final review (and hopefully approval).

**Review by mophahr — APPROVED** [review]
:+1: Thanks fo addressing my concerns :)

**Comment by mophahr** [comment]
@ericzolf will add a link to the Tower docs but apart from than that, this PR was approved in the CoP meeting.

**Comment by ericzolf** [comment]
I'll add a link https://docs.ansible.com/automation-controller/latest/html/userguide/workflow_templates.html#ug-wf-templates-extravars

**Comment by ericzolf** [comment]
Merged as it was previously approved.

**Pull request merged**